File: Sample/Koudai/Server/src/ZyGames.Tianjiexing.BLL/Action/Action1019.cs
/****************************************************************************
Copyright (c) 2013-2015 scutgame.com

http://www.scutgame.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System;
using System.Collections.Generic;
using System.Web;
using System.Web.Caching;
using ZyGames.Framework.Game.Com.Rank;
using ZyGames.Framework.Game.Service;
using ZyGames.Framework.Collection;
using ZyGames.Framework.Common;
using ZyGames.Framework.SyncThreading;
using ZyGames.Tianjiexing.Model;
using ZyGames.Tianjiexing.Model.Config;
using ZyGames.Framework.Game.Com;
using ZyGames.Tianjiexing.Component;


namespace ZyGames.Tianjiexing.BLL.Action
{

    /// <summary>
    /// 1019_排行榜列表接口
    /// </summary>
    public class Action1019 : BaseAction
    {
        private RankType rankType;
        private int pageIndex = 0;
        private int pageSize = 0;
        private int pageCount = 0;
        private int baseNum;
        private IList<UserRank> userRankArray;
        private int currRankID = 0;

        public Action1019(ZyGames.Framework.Game.Contract.HttpGet httpGet)
            : base(ActionIDDefine.Cst_Action1019, httpGet)
        {

        }

        public override void BuildPacket()
        {
            PushIntoStack(pageCount);
            PushIntoStack(userRankArray.Count);
            foreach (UserRank userInfo in userRankArray)
            {
                if (rankType == RankType.UserLv)
                {
                    baseNum = userInfo.UserLv;
                }
                else if (rankType == RankType.Obtion)
                {
                    baseNum = userInfo.ObtainNum;
                }
                else if (rankType == RankType.GameCoin)
                {
                    baseNum = userInfo.GameCoin;
                }
                else if (rankType == RankType.ZhanLi)
                {
                    baseNum = userInfo.TotalCombatNum;
                }
                DataStruct dsItem = new DataStruct();
                dsItem.PushIntoStack(userInfo.RankId);
                dsItem.PushIntoStack(userInfo.UserID.ToNotNullString());
                dsItem.PushIntoStack(userInfo.NickName.ToNotNullString());
                dsItem.PushIntoStack(baseNum);
                dsItem.PushIntoStack((short)userInfo.CountryID);
                dsItem.PushIntoStack((short)userInfo.VipLv);
                PushIntoStack(dsItem);
            }
            PushIntoStack(currRankID);
        }

        public override bool GetUrlElement()
        {
            if (httpGet.GetEnum("RankType", ref rankType)
                 && httpGet.GetInt("PageIndex", ref pageIndex)
                 && httpGet.GetInt("PageSize", ref pageSize))
            {
                return true;
            }
            return false;

        }

        public override bool TakeAction()
        {
            Ranking<UserRank> rankList;
            if (rankType == RankType.UserLv)
            {
                rankList = RankingFactory.Get<UserRank>(UserLvRanking.RankingKey);
            }
            else if (rankType == RankType.Obtion)
            {
                rankList = RankingFactory.Get<UserRank>(ObtainRanking.RankingKey);
            }
            else if (rankType == RankType.GameCoin)
            {
                rankList = RankingFactory.Get<UserRank>(GameCoinRanking.RankingKey);
            }
            else if (rankType == RankType.ZhanLi)
            {
                rankList = RankingFactory.Get<UserRank>(CombatNumRanking.RankingKey);
            }
            else
            {
                throw new ArgumentNullException("rankType", string.Format("rankType:{0}", rankType));
            }
            userRankArray = rankList.GetRange(pageIndex, pageSize, out pageCount);
            if (rankList.TryGetRankNo(m => m.UserID == ContextUser.UserID, out currRankID))
            {

            }
            return true;
        }
    }
}

File: Sample/Doudizhu/Server/src/ZyGames.Doudizhu/Script/PyScript/Action/Action1019.py
import ReferenceLib
from action import *
from System import *
from mathUtils import MathUtils

from System.Collections.Generic import *
from ZyGames.Framework.SyncThreading import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Game.Cache import *
from ZyGames.Framework.Game.Com.Rank import *
from ZyGames.Framework.Game.Service import *
from ZyGames.Doudizhu.Bll import *
from ZyGames.Framework.Game.Lang import *
from ZyGames.Doudizhu.Model import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Framework.Game.Runtime import *
from ZyGames.Framework.Cache import *


#排行榜接口
class UrlParam(HttpParam):
    def __init__(self):
        HttpParam.__init__(self)
        self.RankType = 0;
        self.PageIndex = 0
        self.PageSize = 0


class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)
        self.userRankList = List[UserRank]
        self.PageCount = 0

def getUrlElement(httpGet, parent):
    urlParam = UrlParam();
    if httpGet.Contains("RankType"):
        urlParam.RankType = httpGet.GetEnum[RankType]("RankType")
        urlParam.Result = True;
    else:
        urlParam.Result = False;
    return urlParam;

def takeAction(urlParam, parent):
    actionResult = ActionResult();

    ranking = Ranking[UserRank];
    if urlParam.RankType == RankType.GameCoin:    
        ranking = RankingFactory.Get[UserRank](BeansRanking.RankingKey)
    if urlParam.RankType == RankType.Wining:
        ranking = RankingFactory.Get[UserRank](WinRanking.RankingKey)
    result = ranking.GetRange(urlParam.PageIndex, urlParam.PageSize); 
    if result:
        actionResult.userRankList = result[0]
        actionResult.PageCount = result[1]
    #需要实现
    return actionResult;

def buildPacket(writer, urlParam, actionResult):
    writer.PushIntoStack(actionResult.PageCount)
    writer.PushIntoStack(len(actionResult.userRankList))
    for info in actionResult.userRankList:
        wining = PythonHelper.TransformString(info.Wining)
        DsItem = DataStruct()
        DsItem.PushIntoStack(info.RankId)
        DsItem.PushIntoStack(info.UserID)
        DsItem.PushIntoStack(MathUtils.ToNotNullString(info.NickName))
        DsItem.PushIntoStack(info.GameCoin)
        DsItem.PushIntoStack(MathUtils.ToNotNullString(wining))
        writer.PushIntoStack(DsItem)
    return True;